### PR TITLE
Add demo of enforcing EC in a pipeline

### PR DESCRIPTION
**Pull request by lcarva**
Signed-off-by: Luiz Carvalho <lucarval@redhat.com>

### CODE REVIEW

File: hack/chains/release-pipeline-with-ec-demo.sh [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+" | oc apply -f - > /dev/null
+
+oc describe pipeline simple-release
+
+
+title "Verify Push Secret"
+
+# Check if required secret exists
+oc get secret release-demo > /dev/null
```
Comment (simonbaird) [inline hack/chains/release-pipeline-with-ec-demo.sh:104]: This just makes sure it fails if secret doesn't exist, right?

File: hack/chains/release-pipeline-with-ec-demo.sh [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+echo "The 'pipeline' ServiceAccount has access to these registries:"
+oc get secret -o json \
+    $(oc get sa pipeline -o json | jq '.secrets[].name' -r | xargs) | \
+    jq '.items[] | .data | .".dockercfg" // .".dockerconfigjson" // ""? |
+        @base64d | fromjson | .auths // . | keys[] | " * "+.' -r | sort -u
```
Comment (simonbaird) [inline hack/chains/release-pipeline-with-ec-demo.sh:109]: That's some deep jq magic, I'll trust you on this one.. :smile:

File: hack/chains/release-pipeline-with-ec-demo.sh (hunk 1 above)
Comment (lcarva) [reply hack/chains/release-pipeline-with-ec-demo.sh:104]: Yup. I tried to express that in the comment. Should I try rephrasing it?

**Comment by simonbaird** [comment]
Works nicely afaict. Some demos:
* http://pastebin.test.redhat.com/1047227
* http://pastebin.test.redhat.com/1047234

**Comment by simonbaird** [comment]
Notes/comments:
* Echoing the whole pipeline run definition is quite verbose. That's okay, but maybe you could prune out the managed fields if you want to make it a little neater.
* To make my build pass I needed to modify POLICY_REPO and POLICY_REPO_REF, and I had to go look in the build-definitions to remember what the params are. Maybe you could do some commented out params to help show what they are and make various demo easier.
* For the secret handling, in pipeline-quay-demo and some others iirc there's a param for the name of the secret, see ` --param PUSH_SECRET_NAME` and how it works. That might be more user friendly for those who downloaded a secret file in yaml from quay.io and did an `oc apply my-quay-secret.yaml`. (Just an idea, not a blocker.)

```

```

File: hack/chains/release-pipeline-with-ec-demo.sh [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+#
+# Pre-requisites:
+#
+#   1. Create a Secret named "release-demo" with access to push to the dst-image-ref, and
+#      link it to the pipeline ServiceAccount (so skopeo can copy the image there):
+#      oc create secret docker-registry release-demo quay.io --docker-username=$USER --docker-password=top-secret
+#      oc secrets link pipeline release-demo --for=pull,mount
+#
+#   2. Create a Secret named "cosign-public-key" with the "cosign.pub" attribute to house the
+#      public key to be used for verifying the signature of the image and its attestation.
```
Comment (simonbaird) [inline hack/chains/release-pipeline-with-ec-demo.sh:14]: Would be nice to add that helper command from the build-definitions repo here too. I gotta go look for it now.

Comment (lcarva) [reply hack/chains/release-pipeline-with-ec-demo.sh]: Good idea, done!

**Comment by lcarva** [comment]
>
>     * Echoing the whole pipeline run definition is quite verbose. That's okay, but maybe you could prune out the managed fields if you want to make it a little neater.
>

Agree. I changed it to display just the `.spec` portion of the Pipeline. I used yq which has the added benefit of adding pretty colors to the output.

>     * To make my build pass I needed to modify POLICY_REPO and POLICY_REPO_REF, and I had to go look in the build-definitions to remember what the params are. Maybe you could do some commented out params to help show what they are and make various demo easier.

Good point. Added a couple more.

>     * For the secret handling, in pipeline-quay-demo and some others iirc there's a param for the name of the secret, see ` --param PUSH_SECRET_NAME` and how it works. That might be more user friendly for those who downloaded a secret file in yaml from quay.io and did an `oc apply my-quay-secret.yaml`. (Just an idea, not a blocker.)

I don't think I can control this. The built-in `skopeo-copy` task doesn't take this parameter. It relies on the the secrets linked to the service account running the pod. It does seem like a more opaque approach though.

**Comment by simonbaird** [comment]
Lgtm.

**Review by simonbaird — APPROVED** [review]

**Comment by simonbaird** [comment]
I don't think the prow e2e is useful since this is just a bash script in hack/chains. Let's merge.

**Pull request merged**